internal/pkg/cli/init.go

```diff
@@ -7,122 +7,162 @@ package cli
+	// we have to resort to type-casting the interface. These pointers simplify data access.
+	appType        *string
+	appName        *string
+	dockerfilePath *string
+
+	// Interfaces for dependencies.
 	projStore   archer.ProjectStore
 	envStore    archer.EnvironmentStore
 	envDeployer archer.EnvironmentDeployer
+	ws          archer.Workspace
+	prog        progress
+	prompt      prompter
+
+	promptForShouldDeploy bool // true means that the user set the ShouldDeploy flag explicitly.
+	existingProjects      []string
+}
+
+func NewInitOpts() (*InitOpts, error) {
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, err
+	}
+	ssm, err := ssm.NewStore()
+	if err != nil {
+		return nil, err
+	}
+	sess, err := session.Default()
+	if err != nil {
+		return nil, err
+	}
+	prompt := prompt.New()
+	spin := spinner.New()
+
+	initApp := &InitAppOpts{
+		fs:             &afero.Afero{Fs: afero.NewOsFs()},
+		manifestWriter: ws,
+		prompt:         prompt,
+	}
+
+	return &InitOpts{
+		initApp: initApp,
 
-	ws               archer.Workspace
-	existingProjects []string
+		appType:        &initApp.AppType,
+		appName:        &initApp.AppName,
+		dockerfilePath: &initApp.DockerfilePath,
 
-	prog     progress
-	prompter prompter
+		// TODO remove these dependencies from InitOpts after https://github.com/aws/amazon-ecs-cli-v2/issues/109
+		projStore:   ssm,
+		envStore:    ssm,
+		ws:          ws,
+		envDeployer: cloudformation.New(sess),
+		prompt:      prompt,
+		prog:        spin,
+	}, nil
 }
 
-// Prepare loads contextual data such as any existing projects, the current workspace, etc.
-func (opts *InitOpts) Prepare() {
+func (opts *InitOpts) Run() error {
 	log.Warningln("It's best to run this command in the root of your workspace.")
 	log.Infoln(`Welcome the the ECS CLI! We're going to walk you through some questions to help you get set up
 with a project on ECS. A project is a collection of containerized applications (or micro-services)
 that operate together.` + "\n")
 
-	// If there's a local project, we'll use that and just skip the project question.
-	// Otherwise, we'll load a list of existing projects that the customer can select from.
+	if err := opts.loadProject(); err != nil {
+		return err
+	}
+	if err := opts.loadApp(); err != nil {
+		return err
+	}
+
+	log.Infof("Ok great, we'll set up a %s named %s in project %s.\n",
+		color.HighlightUserInput(*opts.appType), color.HighlightUserInput(*opts.appName), color.HighlightUserInput(opts.Project))
+
+	if err := opts.createProject(); err != nil {
+		return err
+	}
+	if err := opts.ws.Create(opts.Project); err != nil {
+		return fmt.Errorf("create workspace for project %s: %w", opts.Project, err)
+	}
+	if err := opts.initApp.Execute(); err != nil {
+		return fmt.Errorf("execute app init: %w", err)
+	}
+	return opts.deploy()
+}
+
+func (opts *InitOpts) loadProject() (err error) {
+	defer func() {
```

Comment:
I think I'd probably do this:

1. If they're in a WS, use that project name (if they also provided a flag, warn them that we'll ignore it)
2. If they provided it via a flag, run validate.
3. If they select an existing one, we'll let the askProjectName do the validation.

Reply:
Are you saying that named returns and a deferred function is confusing???

Done! I'll have to copy this logic again in https://github.com/aws/amazon-ecs-cli-v2/pull/227 >_>